File: android/app/src/main/java/com/danish/m2m/ChimeVideoPlatformView.kt
package com.danish.m2m

import android.content.Context
import android.view.View
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.DefaultVideoRenderView
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class ChimeVideoPlatformViewFactory(private val plugin: ChimeMeetingPlugin) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val creationParams = args as? Map<String, Any>
        return ChimeVideoPlatformView(context, plugin, creationParams)
    }
}

class ChimeVideoPlatformView(
    context: Context,
    private val plugin: ChimeMeetingPlugin,
    creationParams: Map<String, Any>?
) : PlatformView {
    private val videoRenderView: DefaultVideoRenderView = DefaultVideoRenderView(context)
    private val isLocal: Boolean = creationParams?.get("isLocal") as? Boolean ?: false

    init {
        plugin.registerVideoView(this, isLocal)
    }

    fun getVideoRenderView(): DefaultVideoRenderView {
        return videoRenderView
    }

    override fun getView(): View {
        return videoRenderView
    }

    override fun dispose() {
        plugin.unregisterVideoView(this)
    }
}

File: ios/Runner/ChimeVideoPlatformView.swift
import Flutter
import UIKit
import AmazonChimeSDK

class ChimeVideoPlatformViewFactory: NSObject, FlutterPlatformViewFactory {
    private let plugin: ChimeMeetingPlugin

    init(plugin: ChimeMeetingPlugin) {
        self.plugin = plugin
        super.init()
    }

    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        let creationParams = args as? [String: Any]
        return ChimeVideoPlatformView(frame: frame, plugin: plugin, creationParams: creationParams)
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class ChimeVideoPlatformView: NSObject, FlutterPlatformView {
    private let videoRenderView: DefaultVideoRenderView
    private let plugin: ChimeMeetingPlugin
    private let isLocal: Bool

    init(frame: CGRect, plugin: ChimeMeetingPlugin, creationParams: [String: Any]?) {
        self.videoRenderView = DefaultVideoRenderView(frame: frame)
        self.plugin = plugin
        self.isLocal = creationParams?["isLocal"] as? Bool ?? false
        super.init()

        plugin.registerVideoView(view: self, isLocal: isLocal)
    }

    func view() -> UIView {
        return videoRenderView
    }

    func getVideoRenderView() -> DefaultVideoRenderView {
        return videoRenderView
    }

    deinit {
        plugin.unregisterVideoView(view: self)
    }
}
